[Modules] No transitive source location change

ChuanqiXu9 opened the pull request:
This is part of "no transitive change" patch series, "no transitive source location change". I talked this with @Bigcheese in the tokyo's WG21 meeting.

The idea comes from @jyknight posted on LLVM discourse. That for:

```
// A.cppm
export module A;
...

// B.cppm
export module B;
import A;
...

//--- C.cppm
export module C;
import C;
```

Almost every time A.cppm changes, we need to recompile `B`. Due to we think the source location is significant to the semantics.  But it may be good if we can avoid recompiling `C` if the change from `A` wouldn't change the BMI of B.

# Motivation Example

This patch only cares source locations. So let's focus on source location's example. We can see the full example from the attached test.

```
//--- A.cppm
export module A;
export template <class T>
struct C {
    T func() {
        return T(43);
    }
};
export int funcA() {
    return 43;
}

//--- A.v1.cppm
export module A;

export template <class T>
struct C {
    T func() {
        return T(43);
    }
};
export int funcA() {
    return 43;
}

//--- B.cppm
export module B;
import A;

export int funcB() {
    return funcA();
}

//--- C.cppm
export module C;
import A;
export void testD() {
    C<int> c;
    c.func();
}
```

Here the only difference between `A.cppm` and `A.v1.cppm` is that `A.v1.cppm` has an additional blank line. Then the test shows that two BMI of `B.cppm`, one specified `-fmodule-file=A=A.pcm` and the other specified `-fmodule-file=A=A.v1.pcm`, should have the bit-wise same contents. 

However, it is a different story for C, since C instantiates templates from A, and the instantiation records the source information from module A, which is different from `A` and `A.v1`, so it is expected that the BMI `C.pcm` and `C.v1.pcm` can and should differ.

# Internal perspective of status quo

To fully understand the patch, we need to understand how we encodes source locations and how we serialize and deserialize them.

For source locations, we encoded them as:

```
|
|
| _____ base offset of an imported module
|
|
|
|_____ base offset of another imported module
|
|
|
|
| ___ 0
```

As the diagram shows, we encode the local (unloaded) source location from 0 to higher bits. And we allocate the space for source locations from the loaded modules from high bits to 0. Then the source locations from the loaded modules will be mapped to our source location space according to the allocated offset.

For example, for,

```
// a.cppm
export module a;
...

// b.cppm
export module b;
import a;
...
```

Assuming the offset of a source location (let's name the location as `S`) in a.cppm is 45 and we will record the value `45` into the BMI `a.pcm`. Then in b.cppm, when we import a, the source manager will allocate a space for module 'a' (according to the recorded number of source locations) as the base offset of module 'a' in the current source location spaces. Let's assume the allocated base offset as 90 in this example.  Then when we want to get the location in the current source location space for `S`, we can get it simply by adding `45` to `90` to `135`. Finally we can get the source location for `S` in module B as `135`.

And when we want to write module `b`, we would also write the source location of `S` as `135` directly in the BMI. And to clarify the location `S` comes from module `a`, we also need to record the base offset of  module `a`, 90 in the BMI of `b`.

Then the problem comes. Since the base offset of module 'a' is computed by the number source locations in module 'a'. In module 'b', the recorded base offset of module 'a' will change every time the number of source locations in module 'a' increase or decrease. In other words, the contents of BMI of B will change every time the number of locations in module 'a' changes. This is pretty sensitive. Almost every change will change the number of locations. So this is the problem this patch want to solve.

Let's continue with the existing design to understand what's going on. Another interesting case is:

```
// c.cppm
export module c;
import whatever;
import a;
import b;
...
```

In `c.cppm`, when we import `a`, we still need to allocate a base location offset for it, let's say the value becomes to `200` somehow. Then when we reach the location `S` recorded in module `b`, we need to translate it into the current source location space. The solution is quite simple, we can get it by `135 + (200 - 90) = 245`. In another word, the offset of a source location in current module can be computed as `Recorded Offset + Base Offset of the its module file - Recorded Base Offset`.

Then we're almost done about how we handle the offset of source locations in serializers.

# The high level design of current patch

From the abstract level, what we want to do  is to remove the hardcoded base offset of imported modules and remain the ability to calculate the source location  in a new module unit. To achieve this, we need to be able to find the module file owning a source location from the encoding of the source location.

So in this patch, for each source location, we will store the local offset of the location and the module file index. For  the above example, in `b.pcm`, the source location of `S` will be recorded as `135` directly. And in the new design, the source location of `S` will be recorded as `<1, 45>`. Here `1` stands for the module file index of `a` in module `b`. And `45` means the offset of `S` to the base offset of module `a`.

So the trade-off here is that, to make the BMI more independent, we need to record more abstract information. And I feel it is worthy. The recompilation problem of modules is really annoying and there are still people complaining this. But if we can make this (including stopping other changes transitively), I think this may be a killer feature for modules. And from @Bigcheese , this should be helpful for clang explicit modules too.

And the benchmarking side, I tested this patch against https://github.com/alibaba/async_simple/tree/CXX20Modules. No significant change on compilation time. The size of .pcm files becomes to 204M from 200M. I think the trade-off is pretty fair.

# Some low level details

I didn't use another slot to record the module file index. I tried to use the higher 32 bits of the existing source location encodings to store that information. This design may be safe. Since we use `unsigned` to store source locations but we use uint64_t in serialization. And generally `unsigned` is 32 bit width in most platforms. So it might not be a safe problem. Since all the bits we used to store the module file index is not used before. So the new encodings may be:

```
   |-----------------------|-----------------------|
   |           A           |         B         | C |

  * A: 32 bit. The index of the module file in the module manager + 1. The +1
          here is necessary since we wish 0 stands for the current module file.
  * B: 31 bit. The offset of the source location to the module file containing it.
  * C: The macro bit. We rotate it to the lowest bit so that we can save some 
          space in case the index of the module file is 0.
```

(The B and C is the existing raw encoding for source locations)

Another reason to reuse the same slot of the source location is to reduce the impact of the patch. Since there are a lot of places assuming we can store and get a source location from a slot. And if I tried to add another slot, a lot of codes breaks. I don't feel it is worhty.

Another impact of this decision is that, the existing small optimizations for encoding source location may be invalided. The key of the optimization is that we can turn large values into small values then we can use VBR6 format to reduce the size. But if we decided to put the module file index into the higher bits, then maybe it simply doesn't work. An example may be the `SourceLocationSequence` optimization. 

This will only affect the size of on-disk .pcm files. I don't expect this impact the speed and memory use of compilations. And seeing my small experiments above, I feel this trade off is worthy.

# Correctness

The mental model for handling source location offsets is not so complex and I believe we can solve it by adding module file index to each stored source location. 

For the practical side, since the source location is pretty sensitive, and the patch can pass all the in-tree tests and a small scale projects, I feel it should be correct.

# Future Plans

I'll continue to work on no transitive decl change and no transitive identifier change (if matters) to achieve the goal to stop the propagation of unnecessary changes. But all of this depends on this patch. Since, clearly, the source locations are the most sensitive thing.

---

The release nots and documentation will be added seperately.

ChuanqiXu9 committed:
[Modules] No transitive source location change

ChuanqiXu9 commented:
BTW, after this patch, with reduced BMI (https://github.com/llvm/llvm-project/pull/85050), we can already do something more interesting than reformating:

```
//--- A.cppm
export module A;
int funcA0();
int funcA1();
export int funcA() {
    return funcA0();
}

//--- A.v1.cppm
export module A;

int funcA0();
int funcA1();
export int funcA() {
    return funcA0() + funcA1();
}

//--- B.cppm
export module B;
import A;

export int funcB() {
    return funcA();
}
```

Now the B.pcm will keep unchanged  with `A.pcm` from `A.cppm` and `A.v1.pcm` from `A.v1.cppm`. We changed the implementation of `funcA()` from `return funcA0();` to `return funcA0() + funcA1();`. And the `B.pcm` can still get the same content.

jyknight commented:
+1 on the high-level plan. Switching from a linear offset to a {local-module-index, offset-within-module} pair sounds great!

jansvoboda11 commented:
By default, `SourceLocation` is 32 bits. One bit is used to distinguish macro expansions. Looking at libc++'s module map, it contains 999 top-level modules at this moment. That's 10 bits just to be able to import the (entire) standard library. That leaves 21 bits, restricting local `SourceLocation` space to 2 MB. This doesn't sound feasible. Did I misunderstand?

ChuanqiXu9 commented:
> By default, `SourceLocation` is 32 bits. One bit is used to distinguish macro expansions. Looking at libc++'s module map, it contains 999 top-level modules at this moment. That's 10 bits just to be able to import the (entire) standard library. That leaves 21 bits, restricting local `SourceLocation` space to 2 MB. This doesn't sound feasible. Did I misunderstand?

Yes, I explained this in `Some low level details` section. The size of source location won't be affected. Since the size of source location is unsigned (practically, it is 32 bits in most platforms). And we use uint64_t as a unit in the serializer. So there are 32 bit not used completely. The plan is to store the module file index in the higher 32 bits and it shouldn't be a safe problem.

The only trade-off I saw about this change is that it may increase the size of **on-disk** .pcm files due to we use VBR6 format to decrease the size of small numbers. But on the one side, we still need to pay for more spaces if we want to use `{local-module-index, offset-within-module} pair` (Thanks for the good name suggestion). On the other hand, from the experiment, it shows the overhead is acceptable.

jansvoboda11 commented:
> Yes, I explained this in Some low level details section. The size of source location won't be affected. Since the size of source location is unsigned (practically, it is 32 bits in most platforms). And we use uint64_t as a unit in the serializer. So there are 32 bit not used completely. The plan is to store the module file index in the higher 32 bits and it shouldn't be a safe problem. Maybe the original wording is not so clear. I've updated it.

Thank you, using 64 bits in the serialization format makes sense! This also means that whenever Clang is configured with 64 bit `SourceLocation`, we should be using 96 bits for serialization: 32 bits for the module file index and 64 bits for the offset itself, correct?

> The only trade-off I saw about this change is that it may increase the size of **on-disk** .pcm files due to we use VBR6 format to decrease the size of small numbers. But on the one side, we still need to pay for more spaces if we want to use `{local-module-index, offset-within-module} pair` (Thanks for the good name suggestion). On the other hand, from the experiment, it shows the overhead is acceptable.

Sorry, I don't quite understand. Are you saying you did or did not try to encode this as two separate 32bit values?

ChuanqiXu9 commented:
> > Yes, I explained this in Some low level details section. The size of source location won't be affected. Since the size of source location is unsigned (practically, it is 32 bits in most platforms). And we use uint64_t as a unit in the serializer. So there are 32 bit not used completely. The plan is to store the module file index in the higher 32 bits and it shouldn't be a safe problem. Maybe the original wording is not so clear. I've updated it.
> 
> Thank you, using 64 bits in the serialization format makes sense! This also means that whenever Clang is configured with 64 bit `SourceLocation`, we should be using 96 bits for serialization: 32 bits for the module file index and 64 bits for the offset itself, correct?

If Clang is configured with 64 bit `SourceLocation`, we can't use 96 bits for serialization. We can at most use 64 bits for a slot. In that case, we can only assume the offset of source location **in its own module** (not the global offset!) is not large than 2^32. I hope this may not be true.

> 
> > The only trade-off I saw about this change is that it may increase the size of **on-disk** .pcm files due to we use VBR6 format to decrease the size of small numbers. But on the one side, we still need to pay for more spaces if we want to use `{local-module-index, offset-within-module} pair` (Thanks for the good name suggestion). On the other hand, from the experiment, it shows the overhead is acceptable.
> 
> Sorry, I don't quite understand. Are you saying you did or did not try to encode this as two separate 32bit values?

I **tried** to encode this as two separate 32bit values. But it will break too many codes. Since a lot of places assume that we can encode the source location as an uint64_t.

What I mean is, with VBR6 format (https://llvm.org/docs/BitCodeFormat.html#variable-width-integer),  we can save more space for small integers in **on-disk** .pcm files (the memory representation should be the same). For example, for a 64 bits unsigned int `1`, VBR6 can use only 6 bits to store that `000001` to represent the 64 bits value `1` in the on-disk representations. So that even if I don't use more slots to store the module file index, the size of the .pcm files will increase after all.

jansvoboda11 commented:
> If Clang is configured with 64 bit `SourceLocation`, we can't use 96 bits for serialization. We can at most use 64 bits for a slot. In that case, we can only assume the offset of source location **in its own module** (not the global offset!) is not large than 2^32. I feel this assumption may be valid in a lot of places.
> 
> Or otherwise we can use less bits for module file index (32 bits seems to be too much honestly), then we can use higher 16 bits to store the module file index, and leave the lower 48 bits to store the source location. In this case, the assumption becomes to "the offset of the source location may not large than 2^48". But it is slightly hard to believe we can reach such extreme cases.

Let's see if @statham-arm (who introduced the `SourceLocation::[U]IntTy` typedefs) wants to weight in here.

> I **tried** to encode this as two separate 32bit values. But it will break too many codes. Since a lot of places assume that we can encode the source location as an uint64_t.
> 
> What I mean is, with VBR6 format (https://llvm.org/docs/BitCodeFormat.html#variable-width-integer), we can save more space for small integers in **on-disk** .pcm files (the memory representation should be the same). For example, for a 64 bits unsigned int `1`, VBR6 can use only 6 bits to store that `000001` to represent the 64 bits value `1` in the on-disk representations. So that even if I don't use more slots to store the module file index, the size of the .pcm files will increase after all.

Right. My thinking was that single 64bit value with the module file index in the upper 32 bits would basically disable VBR6 encoding for the lower 32 bits. If we split this thing into two separate 32 bit values, we are more likely to VBR6 encode both of them. But this might actually increase size for (what I assume is the most common case) local source locations. Still, I think having a rough idea of how alternative implementations compare would be great.

Do you have any data on how much recompilation this can save for real world projects?

ChuanqiXu9 commented:
> > If Clang is configured with 64 bit `SourceLocation`, we can't use 96 bits for serialization. We can at most use 64 bits for a slot. In that case, we can only assume the offset of source location **in its own module** (not the global offset!) is not large than 2^32. I feel this assumption may be valid in a lot of places.
> > Or otherwise we can use less bits for module file index (32 bits seems to be too much honestly), then we can use higher 16 bits to store the module file index, and leave the lower 48 bits to store the source location. In this case, the assumption becomes to "the offset of the source location may not large than 2^48". But it is slightly hard to believe we can reach such extreme cases.
> 
> Let's see if @statham-arm (who introduced the `SourceLocation::[U]IntTy` typedefs) wants to weight in here.
> 
> > I **tried** to encode this as two separate 32bit values. But it will break too many codes. Since a lot of places assume that we can encode the source location as an uint64_t.
> > What I mean is, with VBR6 format (https://llvm.org/docs/BitCodeFormat.html#variable-width-integer), we can save more space for small integers in **on-disk** .pcm files (the memory representation should be the same). For example, for a 64 bits unsigned int `1`, VBR6 can use only 6 bits to store that `000001` to represent the 64 bits value `1` in the on-disk representations. So that even if I don't use more slots to store the module file index, the size of the .pcm files will increase after all.
> 
> Right. My thinking was that single 64bit value with the module file index in the upper 32 bits would basically disable VBR6 encoding for the lower 32 bits. If we split this thing into two separate 32bit values, we are more likely to VBR6 encode both of them. But this would actually increase size for (what I assume is the most common case) local source locations. 

Yes, this is the trade offs.

> Still, I think having a rough idea of how alternative implementations compare would be great.
> 
> Do you have any data on how much recompilation this can save for real world projects?

I don't have a specific data though. But I think it is understandable that the feature is super important. And we can save countless unnecessary compilations after this feature.

After the patch, for reduced BMI, we're already able to avoid changing the BMI if we only change the definitions of non-inline function bodies in the module unit.

Further more, based on this patch, we can do further optimizations. e.g., if we split declaration ID like we did in this patch, we may be able to avoid recompilations if a unreferenced module unit adds or deletes declarations. This may be pretty common in practice:

```
export module interface:part1;
...

//--- 
export module interface:part2;
...

//--- export module interface;
export import :part1;
export import :part2;

//--- consumer.cppm
export module consumer;
import interface;
// only used declarations in interface:part1;
```

Then if the user only changes module unit `interface:part2`, then we can keep the BMI for `consumer.cppm` the same. That implies every user of `consumer.cppm` can avoid recompilations.

---

I see this is a killer feature for C++20 modules. I think it is significantly valuable. So I really want to make this (including later optimizations) into clang19.

statham-arm commented:
> Let's see if @statham-arm (who introduced the `SourceLocation::[U]IntTy` typedefs) wants to weight in here.

I'm afraid my knowledge of C++ modules is very close to zero. They were mentioned in a training course I did last year, but not in much detail.

On 64-bit SourceLocation in general: our patch series to implement those as an option in clang was never fully landed, because the second half of it stalled in review. I'd still like to see it finished off, though I'm sure it would need some vigorous rebasing and retesting by now. The 32-bit SourceLocation limit is a problem for at least some of our users, apparently because there's a library of header files that break the limit all by themselves. (I'm not sure how; I haven't seen them. Maybe by including each other multiple times with different `#define`s?)

But I have no idea whether the same considerations would apply to modules, because I don't really know enough about modules, sorry!

ChuanqiXu9 commented:
> > Let's see if @statham-arm (who introduced the `SourceLocation::[U]IntTy` typedefs) wants to weight in here.
> 
> I'm afraid my knowledge of C++ modules is very close to zero. They were mentioned in a training course I did last year, but not in much detail.
> 
> On 64-bit SourceLocation in general: our patch series to implement those as an option in clang was never fully landed, because the second half of it stalled in review. I'd still like to see it finished off, though I'm sure it would need some vigorous rebasing and retesting by now. The 32-bit SourceLocation limit is a problem for at least some of our users, apparently because there's a library of header files that break the limit all by themselves. (I'm not sure how; I haven't seen them. Maybe by including each other multiple times with different `#define`s?)
> 
> But I have no idea whether the same considerations would apply to modules, because I don't really know enough about modules, sorry!

Hi Statham, thanks for quick reply. I guess the abstract question beyond modules may be: is it safe to use 48 bit to store the source location? Then we can use the higher 16 bit to represent module specific informations, e.g., module file index.

statham-arm commented:
Yes, we think that should be safe – we don't think any of our users is generating 2^48 bytes of preprocessed output. Parsing it afterwards would take prohibitively long if they did! Probably they aren't breaking the 2^32 barrier by _much_.

ChuanqiXu9 commented:
> Yes, we think that should be safe – we don't think any of our users is generating 2^48 bytes of preprocessed output. Parsing it afterwards would take prohibitively long if they did! Probably they aren't breaking the 2^32 barrier by _much_.

Thanks. And that makes sense. I never heard compilation crash due to too long source fils before : )

ChuanqiXu9 commented:
Rebase with main and enabling the sequence optimization when the module file index is 0. Now the size of the PCMs goes to 204M (originally it is 200M). It looks better.

@jansvoboda11 @Bigcheese ping

Bigcheese reviewed (commented):
I think the general approach makes sense. I'll take a closer look at the specific changes.

jansvoboda11 commented on clang/include/clang/Serialization/SourceLocationEncoding.h:
```diff
@@ -149,14 +157,44 @@ class SourceLocationSequence::State {
+  assert(Loc.getOffset() >= BaseOffset);
+  Loc = Loc.getLocWithOffset(-BaseOffset);
+  RawLocEncoding Encoded = encodeRaw(Loc.getRawEncoding());
+  assert(Encoded < ((RawLocEncoding)1 << 32));
```
My compiler says:
```
warning: result of comparison of constant 4294967296 with expression of type 'unsigned int' is always true [-Wtautological-constant-out-of-range-compare]
```

ChuanqiXu9 replied:
I guess you may refer to the next line and that one is indeed comparing unsigned with `(1<<32)`, and I fixed that. This line should be fine since Encoded has type `RawLocEncoding `.

jansvoboda11 replied:
No, I'm referring to this line. I assume the compiler sees through `RawLocEncoding` and knows it's being initialized from `SourceLocation::UIntTy`.

ChuanqiXu9 replied:
Oh, the compiler may be too smart : )

My intention was like, if someday we turned 64 bit source location on, this assertion can make us remember  to update here.

But given the warning, I don't know how to write the assertion here. So I just removed it.

jansvoboda11 commented on clang/include/clang/Serialization/ASTReader.h:
```diff
@@ -942,6 +942,12 @@ class ASTReader
+  /// The module files imported by different module files. Indirectly imported
+  /// module files are included too. The information comes from
+  /// ReadModuleOffsetMap(ModuleFile&).
+  mutable llvm::DenseMap<ModuleFile *, llvm::SmallVector<ModuleFile *>>
+      ImportedModuleFiles;
```
Why does this live in `ASTReader`? If we moved it to `ModuleFile`, we could remove `mutable` and the map lookups.

ChuanqiXu9 replied:
Done in the new commit.

There was a similar member `Imports` in `ModuleFile`, which records the directly imported module files from the coding. But, yes, it is good to remove the map lookups and let's try to handle the ambiguous by commenting.

jansvoboda11 commented on clang/lib/Serialization/ASTWriter.cpp:
```diff
@@ -5574,10 +5577,34 @@ void ASTWriter::AddFileID(FileID FID, RecordDataImpl &Record) {
+    // 0 means the location is not loaded. So we need to add 1 to the index to
+    // make it clear.
+    ModuleFileIndex = F->Index + 1;
+    assert(&getChain()->getModuleManager()[F->Index] == F);
```
This seems a bit tautological.

ChuanqiXu9 replied:
Yes but I'd like to remain it. The assertions make me feel better.

jansvoboda11 replied:
I don't think `ASTWriter` should be re-checking `ModuleManager` invariants either way.

jansvoboda11 commented on clang/include/clang/Serialization/ASTReader.h:
```diff
@@ -2220,40 +2227,47 @@ class ASTReader
-                                    LocSeq *Seq = nullptr) const {
-    SourceLocation Loc = ReadUntranslatedSourceLocation(Raw, Seq);
-    return TranslateSourceLocation(ModuleFile, Loc);
+  SourceLocation ReadRawSourceLocation(ModuleFile &MF, RawLocEncoding Raw,
```
Can we undo the rename? This is named after the return type (what the client expects) not after the argument (the raw location). Also, `ASTReader` will never read a non-"raw" source location.

ChuanqiXu9 replied:
Done.

jansvoboda11 commented on clang/include/clang/Serialization/ASTReader.h:
```diff
@@ -696,7 +696,7 @@ class ASTReader
   /// submodule resides along with the offset that should be added to the
   /// global submodule ID to produce a local ID.
-  GlobalSubmoduleMapType GlobalSubmoduleMap;
+  mutable GlobalSubmoduleMapType GlobalSubmoduleMap;
```
Is `mutable` necessary here?

ChuanqiXu9 replied:
Oh, I don't know why it remains. Maybe I added during the work progress.

jansvoboda11 commented:
I left a couple of initial comments. With the 64-bit `SourceLocation` aspect of this change solved and with just 2% size increase, I think this approach makes sense. Could you share the rest of the "no transitive change" patches, or at least their outline? I'm curious what else goes into making that work.

Bigcheese commented on clang/include/clang/Serialization/ASTReader.h:
```diff
@@ -2220,40 +2227,47 @@ class ASTReader
+    if (SourceMgr.isLoadedSourceLocation(Loc))
+      return Loc;
+
+    return Loc.getLocWithOffset(ModuleFile.SLocEntryBaseOffset - 2);
```
Why the -2 here?

ChuanqiXu9 replied:
There are a lot `- 2` in Serialization and SourceManager. I just tried to follow that. Sadly I didn't find clear comments explaining this. By reading the codes, I think it is trying to skip 2 dummy SLocEntries (the invalid source location and invalid expansion location, see `SourceManager::clearIDTables()`) since they are not sensitive to the context.

Bigcheese commented on clang/include/clang/Serialization/ASTReader.h:
```diff
@@ -2220,40 +2227,47 @@ class ASTReader
+
+    auto [Loc, ModuleFileIndex] = ReadUntranslatedSourceLocation(Raw, Seq);
+    ModuleFile *ModuleFileHomingLoc =
+        ModuleFileIndex ? ImportedModuleFiles[&MF][ModuleFileIndex - 1] : &MF;
```
Should have at least a bounds check assert on the ModuleFileIndex value here.

ChuanqiXu9 replied:
We have a bounds check in the `operator[]` of SmallVector. So I guess it may be fine.

Bigcheese reviewed (commented):
I have a few minor comments on the patch. I want to do some additional perf testing on module scanning perf because I'm a bit concerned about the cost of `ASTWriter::getRawSourceLocationEncoding`.

ChuanqiXu9 commented:
> Could you share the rest of the "no transitive change" patches, or at least their outline? I'm curious what else goes into making that work.

Oh, I haven't started them. The patch in my mind includes two at least:
(1) Add a new block recording the semantical hash value of the current named module.
(2) A similar patch for DeclID to refactor the existing linear offset to {ModuleFileIndex, LocalDeclID} patch.

BTW, with (2), maybe we can refactor the existing super large vector of Decls, `DeclsLoaded`, in the ASTReader into a array of vectors or into a map. But this may not be included in the series of the patches.

The motivation for (1) is that:

```
export module a;
export import b;
...
```

We want the BMI of `a` to change all the time if `b` changes. 

Also for the motivation example, if we changed the definition of `funcB` into `inline`,

```
//--- A.cppm
export module A;
export inline int funcA() {
    return 43;
}

//--- B.cppm
export module B;
import A;

export inline int funcB() {
    return funcA();
}
```

We hope the BMI of module B will change after the function A changes. But the declaration level hash may be tricky, so probably we need to change the BMI of the module B after the module A changes as the initial step.

So the result may be, we need to add a new hash or signature value to the BMI for named modules:

```
signature = hash_combine(hash value of the current buffer, combined hash value of the signature export imported modules, combined hash value of touched non-export import modules)
```

(Maybe we need to do this for all module types)

Maybe we can reuse the existing ASTSignature or we need to add a new block. But I am sure about this point now.

ChuanqiXu9 commented on clang/include/clang/Serialization/SourceLocationEncoding.h:
```diff
@@ -149,14 +157,44 @@ class SourceLocationSequence::State {
+  assert(Encoded < ((RawLocEncoding)1 << 32));
+
+  assert(BaseModuleFileIndex < ((RawLocEncoding)1 << 32));
+  Encoded |= (RawLocEncoding)BaseModuleFileIndex << 32;
```
Now I assert it should be less than `(1<<16)`

ChuanqiXu9 commented on clang/include/clang/Serialization/SourceLocationEncoding.h:
```diff
@@ -149,14 +157,44 @@ class SourceLocationSequence::State {
+             : SourceLocation::getFromRawEncoding(decodeRaw(Encoded)),
+            ModuleFileIndex};
+
+  Encoded &= ((RawLocEncoding)1 << 33) - 1;
```
Done.

ChuanqiXu9 commented on clang/include/clang/Serialization/ASTReader.h:
```diff
@@ -2220,40 +2227,47 @@ class ASTReader
+      ReadModuleOffsetMap(MF);
+
+    auto [Loc, ModuleFileIndex] = ReadUntranslatedSourceLocation(Raw, Seq);
+    ModuleFile *ModuleFileHomingLoc =
```
Done.

ChuanqiXu9 commented on clang/lib/Serialization/ASTReader.cpp:
```diff
@@ -4078,8 +4065,8 @@ void ASTReader::ReadModuleOffsetMap(ModuleFile &F) const {
       return;
```
Done.

ChuanqiXu9 commented:
> I have a few minor comments on the patch. I want to do some additional perf testing on module scanning perf because I'm a bit concerned about the cost of `ASTWriter::getRawSourceLocationEncoding`.

What's the perf results about scanning?

jansvoboda11 commented on clang/include/clang/Serialization/ASTReader.h:
```diff
@@ -2220,33 +2221,40 @@ class ASTReader
+
+    auto [Loc, ModuleFileIndex] = ReadUntranslatedSourceLocation(Raw, Seq);
+    ModuleFile *OwningModuleFile =
+        ModuleFileIndex ? MF.DependentModules[ModuleFileIndex - 1] : &MF;
```
Since the type of `ModuleFileIndex` is not obvious due to `auto`, I suggest comparing it to `0` explicitly.

```suggestion
        ModuleFileIndex == 0 ? &MF : MF.DependentModules[ModuleFileIndex - 1];
```

Or maybe the element at index `0` in `DependentModules` could be `MF` itself to simplify this further?

ChuanqiXu9 replied:
Done.

I feel the idea to store `MF` itself at index `0` may be slightly tricky.

jansvoboda11 commented on clang/include/clang/Serialization/ASTReader.h:
```diff
@@ -2220,33 +2221,40 @@ class ASTReader
+    if (Loc.isInvalid())
+      return Loc;
+
+    // It implies that the Loc is already translated.
+    if (SourceMgr.isLoadedSourceLocation(Loc))
+      return Loc;
```
How can we get here with a `SourceLocation` that's already been translated?

ChuanqiXu9 replied:
My thought was like "all the recorded source location are local to its module file", so it won't be translated. But after I review how we judge whether a source location is loaded in source manager, I find this may be dangerous in extreme cases.  We judge if a source location is loaded by comparing the offset with the current loaded offset. So it may be dangerous if we have too many loaded source location in the TU and the source location we read is too big, then we may misread it as loaded source location.

My intention for this code is to make `TranslateSourceLocation` to be symmetric with the original one. The original `TranslateSourceLocation` is re-enterable. We could call `TranslateSourceLocation`  on a translated source location. But we can't do this for the `TranslateSourceLocation`   in the current patch.

So I just removed the check and leave a FIXME comment. I'd like to fix this as a NFC patch after we land this.

jansvoboda11 replied:
I don't think this an issue. Now that `TranslateSourceLocation()` is only called from `ReadSourceLocation()` there's no reason for taking care to handle already-translated source locations correctly. Maybe now that it doesn't use any `ASTWriter` members we can make the function a non-member and make it static `ASTWriter.cpp`. In my mind that's enough to leave out the FIXME and just assume it only works when used on an untranslated source location read in `ReadSourceLocation()`.

As an optional clarification we can return a special type (e.g. `UntranslatedSourceLocation`) from `ReadUntranslatedSourceLocation()` and make it so that only `TranslateSourceLocation()` can transform that into the regular (now always-translated) `SourceLocation`.

ChuanqiXu9 replied:
> Now that TranslateSourceLocation() is only called from ReadSourceLocation()

Sadly, this is not true. `TranslateSourceLocation()` may be called in `ASTReader::ReadAST()`:

https://github.com/llvm/llvm-project/blob/aac4d03423dd6b7bdef0f2eb03c570f3e2ca6630/clang/lib/Serialization/ASTReader.cpp#L4588-L4591

The input value of  `TranslateSourceLocation()` there may come from a reading of untranslated source location in `ASTReader::ReadControlBlock` when reading imported modules. Or the input value may come from the argument of  `ASTReader::ReadAST()`, where must be a translated source location. 

Then it looks really dangerous to me. So I add the FIXME. We may not be able to change the signature of the argument of `TranslateSourceLocation()` to `UntranslatedSourceLocation` since that will require us to change the signature of `ASTReader::ReadAST()`.

The reason why actually it works, is that, in the case the input value comes from a translated source location (passed directly in `ASTReader::ReadAST()`), the value of `M.ImportedBy` may always be null **now**.  But I feel it is dangerous if someone changes it suddenly.

jansvoboda11 replied:
I see, thanks for explaining that.

I think we should still attempt to unify things so that `TranslateSourceLocation()` is only called from `ReadSourceLocation()`. Even if it means carrying something like `std::variant<SourceLocation, SourceLocationEncoding::RawLocEncoding>` (or `, UntranslatedSourceLocation>`) in few places. Juggling translated and untranslated `SourceLocation` instances and making functions resilient to both seems unnecessarily complex. I'm fine with leaving a FIXME in this patch and doing a follow-up.

ChuanqiXu9 commented:
Fix conflicts and rebase with main.

@Bigcheese @jansvoboda11  ping~

jansvoboda11 commented on clang/include/clang/Serialization/ASTReader.h:
```diff
@@ -2221,33 +2222,45 @@ class ASTReader
+    ModuleFile *OwningModuleFile =
+        ModuleFileIndex == 0 ? &MF : MF.DependentModules[ModuleFileIndex - 1];
+
+    assert(!SourceMgr.isLoadedSourceLocation(Loc) && "Run out source location space");
```
This is already guarded against in `ReadASTBlock()`, no need to have redundant checks.

ChuanqiXu9 replied:
But the value of `Loc`  may not be valid. Also I feel it is fine to have some redundant assertions. It helps the reader to understand the codes better.

jansvoboda11 replied:
What does `Loc` being invalid has to do with this?

jansvoboda11 commented on clang/lib/Serialization/ASTReader.cpp:
```diff
@@ -4082,14 +4069,14 @@ void ASTReader::ReadModuleOffsetMap(ModuleFile &F) const {
     if (!OM) {
       std::string Msg =
-          "SourceLocation remap refers to unknown module, cannot find ";
+          "cannot find module ";
```
Would be good to keep some context in the error message, maybe remove just the `"SourceLocation"` part and leave the rest as it was?

ChuanqiXu9 replied:
Done

ChuanqiXu9 commented:
@jansvoboda11 @Bigcheese ping

jansvoboda11 reviewed (approved):
The changes LGTM, don't want to block this on my remaining nits.

I believe @Bigcheese wanted to test test impact on PCM size on our side before this lands.

ChuanqiXu9 commented:
> The changes LGTM, don't want to block this on my remaining nits.

Thanks for reviewing this.

> 
> I believe @Bigcheese wanted to test test impact on PCM size on our side before this lands.

I've rebased this with main. I'll wait for the results from @Bigcheese

Bigcheese reviewed (approved):
I did some testing today and this change seems fine. For scanning modules I actually saw some get smaller with your change.

rorth commented:
I strongly suspect that this patch badly broke the [Solaris/sparcv9 buildbot](https://lab.llvm.org/buildbot/#/builders/72/builds/4046): it introduced more than 1000 failures.

Please fix or revert.

ChuanqiXu9 commented:
> I strongly suspect that this patch badly broke the [Solaris/sparcv9 buildbot](https://lab.llvm.org/buildbot/#/builders/72/builds/4046): it introduced more than 1000 failures.
> 
> Please fix or revert.

I'll revert this. Due to I can't reproduce this. When the bot gets stable, please tell if it is the real problem.

rorth commented:
> I'll revert this. Due to I can't reproduce this. When the bot gets stable, please tell if it is the real problem.

You can reproduce this: the [GCC compile farm](https://portal.cfarm.net/) does have a Solaris/sparcv9 system (`cfarm215`) which is perfectly equipped to run LLVM builds (I've tried).

I think the stack traces from the bot are a pretty strong indication that your patch is the culprit:
```
Stack dump without symbol names (ensure you have llvm-symbolizer in your PATH or set the environment var `LLVM_SYMBOLIZER_PATH` to point to it):
0  clang-19  0x00000001076d87b8 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) + 36
1  clang-19  0x00000001076d910c SignalHandler(int) + 896
2  libc.so.1 0x00007fffff0c62a8 __sighndlr + 12
3  libc.so.1 0x00007fffff0b8b50 call_user_handler + 1024
4  libc.so.1 0x00007fffff0b8f10 sigacthandler + 160
5  clang-19  0x00000001083824e0 clang::ASTReader::DeclCursorForID(clang::GlobalDeclID, clang::SourceLocation&) + 168
6  clang-19  0x000000010838aca0 clang::ASTReader::ReadDeclRecord(clang::GlobalDeclID) + 48
7  clang-19  0x00000001082fb4ec clang::ASTReader::GetDecl(clang::GlobalDeclID) + 232
8  clang-19  0x00000001082cb820 clang::ASTReader::SetGloballyVisibleDecls(clang::IdentifierInfo*, llvm::SmallVectorImpl<clang::GlobalDeclID> const&, llvm::SmallVectorImpl<clang::Decl*>*) + 252
9  clang-19  0x00000001083144a0 clang::ASTReader::finishPendingActions() + 572
10 clang-19  0x0000000108319e10 clang::ASTReader::FinishedDeserializing() + 92
11 clang-19  0x000000010830dbf4 clang::ASTReader::get(llvm::StringRef) + 680
12 clang-19  0x00000001078a84fc clang::IdentifierTable::get(llvm::StringRef) + 84
13 clang-19  0x000000010a130fcc clang::Sema::Initialize() + 1208
14 clang-19  0x0000000109fd1814 clang::Parser::Initialize() + 1260
15 clang-19  0x0000000109fccb68 clang::ParseAST(clang::Sema&, bool, bool) + 556
16 clang-19  0x00000001081b10d8 clang::ASTFrontendAction::ExecuteAction() + 248
17 clang-19  0x00000001081b06f8 clang::FrontendAction::Execute() + 92
18 clang-19  0x00000001081196c8 clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) + 1572
19 clang-19  0x00000001082b87b8 clang::ExecuteCompilerInvocation(clang::CompilerInstance*) + 684
20 clang-19  0x00000001048a2980 cc1_main(llvm::ArrayRef<char const*>, char const*, void*) + 4296
21 clang-19  0x000000010489f6f8 ExecuteCC1Tool(llvm::SmallVectorImpl<char const*>&, llvm::ToolContext const&) + 1184
22 clang-19  0x000000010489e018 clang_main(int, char**, llvm::ToolContext const&) + 4424
23 clang-19  0x00000001048aee0c main + 60
24 clang-19  0x000000010489c904 _start + 100
/var/llvm/dist-sparcv9-release-stage2-A-flang-clang18/tools/clang/stage2-bins/tools/clang/test/PCH/Output/opencl-extensions.cl.script: line 2: 12701 Bus Error               /var/llvm/dist-sparcv9-release-stage2-A-flang-clang18/tools/clang/stage2-bins/bin/clang -cc1 -internal-isystem /var/llvm/dist-sparcv9-release-stage2-A-flang-clang18/tools/clang/stage2-bins/lib/clang/19/include -nostdsysteminc -include-pch /var/llvm/dist-sparcv9-release-stage2-A-flang-clang18/tools/clang/stage2-bins/tools/clang/test/PCH/Output/opencl-extensions.cl.tmp -fsyntax-only /vol/llvm/src/llvm-project/dist/clang/test/PCH/opencl-extensions.cl -triple spir-unknown-unknown
```
One thing I see immediately that this uses a triple the bot is not configured to handle.  Nonetheless clang shouldn't die with `SIGBUS` in such as case.

ChuanqiXu9 commented:
> > I'll revert this. Due to I can't reproduce this. When the bot gets stable, please tell if it is the real problem.
> 
> You can reproduce this: the [GCC compile farm](https://portal.cfarm.net/) does have a Solaris/sparcv9 system (`cfarm215`) which is perfectly equipped to run LLVM builds (I've tried).
> 
> I think the stack traces from the bot are a pretty strong indication that your patch is the culprit:
> 
> ```
> Stack dump without symbol names (ensure you have llvm-symbolizer in your PATH or set the environment var `LLVM_SYMBOLIZER_PATH` to point to it):
> 0  clang-19  0x00000001076d87b8 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) + 36
> 1  clang-19  0x00000001076d910c SignalHandler(int) + 896
> 2  libc.so.1 0x00007fffff0c62a8 __sighndlr + 12
> 3  libc.so.1 0x00007fffff0b8b50 call_user_handler + 1024
> 4  libc.so.1 0x00007fffff0b8f10 sigacthandler + 160
> 5  clang-19  0x00000001083824e0 clang::ASTReader::DeclCursorForID(clang::GlobalDeclID, clang::SourceLocation&) + 168
> 6  clang-19  0x000000010838aca0 clang::ASTReader::ReadDeclRecord(clang::GlobalDeclID) + 48
> 7  clang-19  0x00000001082fb4ec clang::ASTReader::GetDecl(clang::GlobalDeclID) + 232
> 8  clang-19  0x00000001082cb820 clang::ASTReader::SetGloballyVisibleDecls(clang::IdentifierInfo*, llvm::SmallVectorImpl<clang::GlobalDeclID> const&, llvm::SmallVectorImpl<clang::Decl*>*) + 252
> 9  clang-19  0x00000001083144a0 clang::ASTReader::finishPendingActions() + 572
> 10 clang-19  0x0000000108319e10 clang::ASTReader::FinishedDeserializing() + 92
> 11 clang-19  0x000000010830dbf4 clang::ASTReader::get(llvm::StringRef) + 680
> 12 clang-19  0x00000001078a84fc clang::IdentifierTable::get(llvm::StringRef) + 84
> 13 clang-19  0x000000010a130fcc clang::Sema::Initialize() + 1208
> 14 clang-19  0x0000000109fd1814 clang::Parser::Initialize() + 1260
> 15 clang-19  0x0000000109fccb68 clang::ParseAST(clang::Sema&, bool, bool) + 556
> 16 clang-19  0x00000001081b10d8 clang::ASTFrontendAction::ExecuteAction() + 248
> 17 clang-19  0x00000001081b06f8 clang::FrontendAction::Execute() + 92
> 18 clang-19  0x00000001081196c8 clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) + 1572
> 19 clang-19  0x00000001082b87b8 clang::ExecuteCompilerInvocation(clang::CompilerInstance*) + 684
> 20 clang-19  0x00000001048a2980 cc1_main(llvm::ArrayRef<char const*>, char const*, void*) + 4296
> 21 clang-19  0x000000010489f6f8 ExecuteCC1Tool(llvm::SmallVectorImpl<char const*>&, llvm::ToolContext const&) + 1184
> 22 clang-19  0x000000010489e018 clang_main(int, char**, llvm::ToolContext const&) + 4424
> 23 clang-19  0x00000001048aee0c main + 60
> 24 clang-19  0x000000010489c904 _start + 100
> /var/llvm/dist-sparcv9-release-stage2-A-flang-clang18/tools/clang/stage2-bins/tools/clang/test/PCH/Output/opencl-extensions.cl.script: line 2: 12701 Bus Error               /var/llvm/dist-sparcv9-release-stage2-A-flang-clang18/tools/clang/stage2-bins/bin/clang -cc1 -internal-isystem /var/llvm/dist-sparcv9-release-stage2-A-flang-clang18/tools/clang/stage2-bins/lib/clang/19/include -nostdsysteminc -include-pch /var/llvm/dist-sparcv9-release-stage2-A-flang-clang18/tools/clang/stage2-bins/tools/clang/test/PCH/Output/opencl-extensions.cl.tmp -fsyntax-only /vol/llvm/src/llvm-project/dist/clang/test/PCH/opencl-extensions.cl -triple spir-unknown-unknown
> ```
> 
> One thing I see immediately that this uses a triple the bot is not configured to handle. Nonetheless clang shouldn't die with `SIGBUS` in such as case.

Reverted. It looks like the configuration isn't in our bots actually. I can't open that site. I need to take another look at the code though.

ChuanqiXu9 commented:
Oh, maybe I found the reason. It is because my patch breaks the alignments of `DeclOffset`: https://github.com/llvm/llvm-project/blob/8d28e5861f8b117a547850ffbb9a332aa6e91459/clang/include/clang/Serialization/ASTBitCodes.h#L237-L240

then it explains why it work well in some platforms but not in other platforms.

rorth commented:
This is certainly a case of unaligned access.  In a local build, I've run the first failing `clang` invocation under `truss` (the Solaris syscall tracer).  For
```
/var/llvm/dist-sparcv9-release-stage2-A-flang-clang18/tools/clang/stage2-bins/bin/clang -cc1 -internal-isystem /var/llvm/dist-sparcv9-release-stage2-A-flang-clang18/tools/clang/stage2-bins/lib/clang/19/include -nostdsysteminc -fmodules -Wno-private-module -fimplicit-module-maps -fmodules-cache-path=/var/llvm/dist-sparcv9-release-stage2-A-flang-clang18/tools/clang/stage2-bins/tools/clang/test/APINotes/Output/availability.m.tmp/ModulesCache -fapinotes-modules -fsyntax-only -I /vol/llvm/src/llvm-project/dist/clang/test/APINotes/Inputs/Headers -F /vol/llvm/src/llvm-project/dist/clang/test/APINotes/Inputs/Frameworks /vol/llvm/src/llvm-project/dist/clang/test/APINotes/availability.m -verify
```
this reveals
```
14552:      Incurred fault #5, FLTACCESS  %pc = 0x1083824E0
14552:        siginfo: SIGBUS BUS_ADRALN addr=0xFFFFFFFF7F5E7B6C
14552:      Received signal #10, SIGBUS [caught]
14552:        siginfo: SIGBUS BUS_ADRALN addr=0xFFFFFFFF7F5E7B6C
```
`gdb` shows exactly that:
```
Thread 2 received signal SIGBUS, Bus error.
[Switching to Thread 1 (LWP 1)]
0x00000001083824e0 in clang::ASTReader::DeclCursorForID(clang::GlobalDeclID, clang::SourceLocation&) ()
1: x/i $pc
=> 0x1083824e0 <_ZN5clang9ASTReader15DeclCursorForIDENS_12GlobalDeclIDERNS_14SourceLocationE+168>:	ldx  [ %i5 + %i1 ], %o2
(gdb) p/x $i5
$1 = 0xffffffff7f5e7b4c
(gdb) p/x $i1
$2 = 0x20
```
The `ldx` insn (Load Extended Word) takes a doubleword address with natural (64-bit) alignment.

ChuanqiXu9 commented:
thanks,  it is pretty helpful.

ChuanqiXu9 commented:
I've relanded this in https://github.com/llvm/llvm-project/commit/947b06282324db8fe2784c4054af9de493a876af.
Let's see what happens.
